components/engine/src/rocks/util/engine_metrics.rs

```diff
@@ -418,471 +443,230 @@ pub fn flush_engine_ticker_metrics(t: TickerType, value: u64, name: &str) {
+            STORE_ENGINE_BLOB_LOCATE_VEC
+                .with_label_values(&[name, "number_blob_next"])
+                .inc_by(v);
+        }
+        TickerType::BlobDbNumPrev => {
+            STORE_ENGINE_BLOB_LOCATE_VEC
+                .with_label_values(&[name, "number_blob_prev"])
+                .inc_by(v);
+        }
+        TickerType::BlobDbNumKeysWritten => {
+            STORE_ENGINE_BLOB_FLOW_VEC
+                .with_label_values(&[name, "keys_written"])
+                .inc_by(v);
+        }
+        TickerType::BlobDbNumKeysRead => {
+            STORE_ENGINE_BLOB_FLOW_VEC
+                .with_label_values(&[name, "keys_read"])
+                .inc_by(v);
+        }
+        TickerType::BlobDbBlobFileBytesWritten => {
+            STORE_ENGINE_BLOB_FLOW_VEC
+                .with_label_values(&[name, "bytes_written"])
+                .inc_by(v);
+        }
+        TickerType::BlobDbBlobFileBytesRead => {
+            STORE_ENGINE_BLOB_FLOW_VEC
+                .with_label_values(&[name, "bytes_read"])
+                .inc_by(v);
+        }
+        TickerType::BlobDbBlobFileSynced => {
+            STORE_ENGINE_BLOB_FILE_SYNCED
+                .with_label_values(&[name, "file_sync"])
+                .inc_by(v);
+        }
+        TickerType::BlobDbGcNumFiles => {
+            STORE_ENGINE_BLOB_GC_FILE_VEC
+                .with_label_values(&[name, "blob_db_num_seek"])
+                .inc_by(v);
+        }
+        TickerType::BlobDbGcNumNewFiles => {
+            STORE_ENGINE_BLOB_GC_FILE_VEC
+                .with_label_values(&[name, "blob_db_num_seek"])
+                .inc_by(v);
+        }
+        TickerType::BlobDbBytesWritten => {
+            STORE_ENGINE_BLOB_GC_FLOW_VEC
+                .with_label_values(&[name, "bytes_written"])
+                .inc_by(v);
+        }
+        TickerType::BlobDbBytesRead => {
+            STORE_ENGINE_BLOB_GC_FLOW_VEC
+                .with_label_values(&[name, "bytes_read"])
+                .inc_by(v);
+        }
+        TickerType::BlobDbGcNumKeysOverwritten => {
+            STORE_ENGINE_BLOB_GC_FLOW_VEC
+                .with_label_values(&[name, "keys_overwritten"])
+                .inc_by(v);
+        }
+        TickerType::BlobDbGcNumKeysRelocated => {
+            STORE_ENGINE_BLOB_GC_FLOW_VEC
+                .with_label_values(&[name, "keys_relocated"])
+                .inc_by(v);
+        }
+        TickerType::BlobDbGcBytesOverwritten => {
+            STORE_ENGINE_BLOB_GC_FLOW_VEC
+                .with_label_values(&[name, "bytes_overwritten"])
+                .inc_by(v);
+        }
+        TickerType::BlobDbGcBytesRelocated => {
+            STORE_ENGINE_BLOB_GC_FLOW_VEC
+                .with_label_values(&[name, "bytes_relocated"])
+                .inc_by(v);
+        }
         _ => {}
     }
 }
 
+macro_rules! engine_histogram_metrics {
+    ($name:ident, $db:expr, $value:expr) => {
+        concat_idents!(STORE_ENGINE_, $name, _VEC)
```

Comment:
Could we *not* use `concat_idents!`? There's no plan to stabilize this macro since the current form is useless (rust-lang/rust#29599), and invocation of `engine_histogram_metrics` itself can be slightly modified to avoid both `concat_idents!` and `casey::lower!`.

```rust
macro_rules! engine_histogram_metrics {
    ($metric:ident, $prefix:expr, $db:expr, $value:expr) => {
        $metric.with_label_values(&[$db, concat!($prefix, "_median")])
            .set($value.median);
        ...
    }
}

...

engine_histogram_metrics!(STORE_ENGINE_GET_MICROS_VEC, "get", name, value);
```

Reply:
The purpose is not to write verbose part, "get" can be deduced by the name of metrics variable. So it is not recommended?

Reply:
Totally not recommended. Unlike other `#![feature]`s used in TiKV, this one is not planned for stabilization until macro hygiene and early macro expansion are both ready, both are complicated features and the latter one doesn't seem actively worked on. So relying on `concat_idents!` will lock TiKV as nightly-only for much longer.

Reply:
@breeswish seem we use this heavily in prometheus, right?